### PR TITLE
Feature/vueify display show

### PR DESCRIPTION
**Pull request by p0psicles**
- [x] PR is based on the DEVELOP branch
- [x] Don't send big changes all at once. Split up big PRs into multiple smaller PRs that are easier to manage and review
- [x] Read the [contribution guide](https://github.com/pymedusa/Medusa/blob/master/.github/CONTRIBUTING.md)

### CODE REVIEW

**Comment by p0psicles** [comment]
please ignore the long list of commits. I branched before it was merged (and sqaushed)

**Comment by p0psicles** [comment]
@medariox @sharkykh fya

File: themes-default/slim/views/displayShow.mako [hunk 1]
```diff
@@ -1,300 +1,187 @@
...
+                                                    <img v-if="config" data-ep-manual-search :src="'images/manualsearch' + (config.themeName === 'dark' ? '-white' : '') + '.png'" width="16" height="16" alt="search" title="Manual Search" />
+                                                </app-link>
+                                            </h3>
+                                            <div class="season-scene-exception" data-season=""></div>
+                                            <div class="pull-right"> <!-- column select and hide/show episodes -->
+                                                    <!-- if not app.DISPLAY_ALL_SEASONS: -->
+                                                    <button v-if="!config.displayAllSeasons" :id="'showseason-' + season.season" type="button" class="btn-medusa pull-right" data-toggle="collapse" :data-target="'#collapseSeason-' + season.season">Hide Episodes</button>
```
Comment (sharkykh) [inline themes-default/slim/views/displayShow.mako]: `config.layout.show.allSeasons`
https://github.com/pymedusa/Medusa/blob/bcec3877de05dbb4ab1ef9fd55202b0d75e1f9d3/medusa/server/api/v2/config.py#L542

File: themes-default/slim/views/displayShow.mako [hunk 2]
```diff
@@ -1,300 +1,187 @@
...
+                                                    <!-- endif -->
+                                                <button id="popover" type="button" class="btn-medusa pull-right selectColumns">Select Columns <b class="caret"></b></button>
+                                            </div> <!-- end column select and hide/show episodes -->
+                                        </th>
+                                    </tr>
+                                </tbody>
+                                <tbody class="tablesorter-no-sort">
+                                    <tr :id="'season-' + season.season + '-cols'" class="seasoncols">
+                                        <th data-column="0" class="col-checkbox"><input type="checkbox" class="seasonCheck" id="" /></th>
```
Comment (sharkykh) [inline themes-default/slim/views/displayShow.mako]: You removed the `id` value by mistake?

File: themes-default/slim/views/displayShow.mako [hunk 3]
```diff
@@ -1,300 +1,187 @@
...
+                                                class="sceneAbsolute form-control input-scene" :data-for-absolute="episode.absoluteNumber || 0"
+                                                :id="'sceneSeasonXEpisode_' + show.id[show.indexer] + episode.absoluteNumber"
+                                                title="Change this value if scene absolute numbering differs from the indexer absolute numbering. Generally used for anime shows."
+                                                :value="getSceneAbsoluteNumbering(episode) ? getSceneAbsoluteNumbering(episode) : ''"
+                                                style="padding: 0; text-align: center; max-width: 60px;"/>
+                                        </td>
+                                        <td class="col-name hidden-xs triggerhighlight">
+                                            <!-- < has_plot = 'has-plot' if epResult['description'] else '' > -->
+                                            <plot-info :has-plot="episode.description ? true : false" :show-slug="indexer + id" :season="String(episode.season)" episode="String(episode.episode)"></plot-info>
```
Comment (sharkykh) [inline themes-default/slim/views/displayShow.mako]: `:has-plot="Boolean(episode.description)"` is more readable IMO

File: themes-default/slim/src/components/show.vue [hunk 4]
```diff
@@ -2,6 +2,7 @@
 import { mapState, mapGetters, mapActions } from 'vuex';
 import { apiRoute } from '../api';
 import { AppLink, PlotInfo } from './helpers';
+import { humanFileSize } from '../utils';
```
Comment (sharkykh) [inline themes-default/slim/src/components/show.vue]: I knew you would need this :wink:

File: themes-default/slim/yarn.lock [hunk 5]
```diff
@@ -3188,6 +3187,10 @@ des.js@^1.0.0:
     inherits "^2.0.1"
     minimalistic-assert "^1.0.0"
 
+desandro-matches-selector@^2.0.0:
```
Comment (sharkykh) [inline themes-default/slim/yarn.lock]: `yarn.lock` shouldn't be modified.
Do you want me to add it to the travis check, so it would fail if you committed an unnecessary change to `yarn.lock`?

File: themes-default/slim/static/css/style.css [hunk 6]
```diff
@@ -1437,7 +1441,7 @@ span.snatched b {
     font-weight: 800;
 }
 
-.displayShowTable {
+.displayShowTable > table {
```
[the remark this answers is not in the document]
Comment (p0psicles) [reply themes-default/slim/static/css/style.css]: Yeah I need to. The tablesorter doesn't work nice with vue. And overwrites classes.

File: themes-default/slim/views/displayShow.mako (hunk 2 above)
Comment (p0psicles) [reply themes-default/slim/views/displayShow.mako]: Yeah hasty coding

File: themes-default/slim/views/displayShow.mako (hunk 3 above)
Comment (p0psicles) [reply themes-default/slim/views/displayShow.mako]: Agree

File: themes-default/slim/static/css/style.css (hunk 6 above)
[the remark this answers is not in the document]
Comment (sharkykh) [reply themes-default/slim/static/css/style.css]: Yup. We knew this was going to be an issue.
And I didn't see anything as good as tablesorter for Vue.

File: themes-default/slim/src/components/show.vue (hunk 4 above)
Comment (p0psicles) [reply themes-default/slim/src/components/show.vue]: Haha. Yeah that was an easy one.

File: themes-default/slim/yarn.lock (hunk 5 above)
Comment (p0psicles) [reply themes-default/slim/yarn.lock]: Okay will do that

File: themes-default/slim/static/css/style.css (hunk 6 above)
[the remark this answers is not in the document]
Comment (p0psicles) [reply themes-default/slim/static/css/style.css]: We can have tablesorter with vue. I'm pretty confident

[the remark this answers is not in the document]
Comment (p0psicles) [reply themes-default/slim/static/css/style.css]: And if not we can implemt it in native vue

[the remark this answers is not in the document]
Comment (medariox) [reply themes-default/slim/static/css/style.css]: Don't know if this is a good option, but https://github.com/xaksis/vue-good-table

File: medusa/server/web/home/handler.py [hunk 7]
```diff
@@ -795,109 +795,15 @@ def displayShow(self, indexername=None, seriesid=None, ):
...
-        )
-
         t = PageTemplate(rh=self, filename='displayShow.mako')
         submenu = [{
             'title': 'Edit',
             'path': 'home/editShow?indexername={series_obj.indexer_name}&seriesid={series_obj.series_id}'.format(series_obj=series_obj),
             'icon': 'ui-icon ui-icon-pencil',
         }]
 
-        if not app.show_queue_scheduler.action.isBeingAdded(series_obj):
```
[the remark this answers is not in the document]
Comment (p0psicles) [reply medusa/server/web/home/handler.py]: i'll revert that for now.

File: medusa/tv/episode.py [hunk 8]
```diff
@@ -1076,6 +1076,8 @@ def to_json(self, detailed=True):
         data['scene'] = {}
         data['scene']['season'] = self.scene_season
         data['scene']['episode'] = self.scene_episode
+        data['hasNfo'] = self.hasnfo
+        data['hasTbn'] = self.hastbn
```
[the remark this answers is not in the document]
Comment (p0psicles) [reply medusa/tv/episode.py]: I like the structure of using data.content here. I'll move hasNfo and hasTbn under data.content.

File: themes-default/slim/static/css/style.css (hunk 6 above)
[the remark this answers is not in the document]
Comment (p0psicles) [reply themes-default/slim/static/css/style.css]: That's basically a component that you pass an array with flat objects. And renders a complete table. I think thats why @sharkykh couldn't find anything as "powerfull" as table sorter.

[the remark this answers is not in the document]
Comment (p0psicles) [reply themes-default/slim/static/css/style.css]: I think i'm going for vuetify data-tables. It's time for some change anyway. And this will give us the opportunity to add cards for mobile.

[the remark this answers is not in the document]
Comment (p0psicles) [reply themes-default/slim/static/css/style.css]: @sharkykh I can't get it to recognize the v-data-table component. Is it maybe because of the way we've hacked in vue?

[the remark this answers is not in the document]
Comment (sharkykh) [reply themes-default/slim/static/css/style.css]: > @sharkykh I can't get it to recognize the v-data-table component. Is it maybe because of the way we've hacked in vue?

Try adding it globally, like `Snotify`, for now:
https://github.com/pymedusa/Medusa/blob/597af4fb4159aecb9f6ba6973cb57568440d1ad6/themes-default/slim/views/layouts/main.mako#L174
https://github.com/pymedusa/Medusa/blob/597af4fb4159aecb9f6ba6973cb57568440d1ad6/themes-default/slim/src/index.js#L65
https://github.com/pymedusa/Medusa/blob/597af4fb4159aecb9f6ba6973cb57568440d1ad6/themes-default/slim/src/app.js#L33

[the remark this answers is not in the document]
Comment (p0psicles) [reply themes-default/slim/static/css/style.css]: I tried this. Didn't work. Doesn't recognize the v-data-table. I can commit it so you can see for yourself.

**Comment by sharkykh** [comment]
Just a general guideline - 
You should plan ahead and think of the show page in terms of components:
- Application
	- AppHeader
	- Show
		- ShowHeader
		- **SeasonTable** (_props:_ indexer, id, seasonId)
			- **EpisodeRow** (_props:_ indexer, id, seasonId, episodeId)
				- (Could have other smaller components here for the "manual search" buttons for example)
	- AppFooter

Then you can use the store to get data, and control stuff using events and props.
It might interfere with your plans to use Vuetify but maybe not, I don't know.
In any case, it will make it should make the code less crowded and maybe more readable..

**Comment by p0psicles** [comment]
Vuetify table supports nested tables. So that was the general idea. The episodes will show as table rows on wide screens and as cards (li) on mobile

**Comment by sharkykh** [comment]
I didn't mean nested tables though.
SeasonTable's main element would be `<table>` and EpisodeRow would be `<tr>`

File: themes-default/slim/static/css/style.css (hunk 6 above)
[the remark this answers is not in the document]
Comment (p0psicles) [reply themes-default/slim/static/css/style.css]: If I can't get vuetify to work I'll try https://github.com/matfish2/vue-tables-2/blob/master/README.md#nested-data
